File: androidApp/core/database/src/main/java/com/grateful/deadly/core/database/AnalyticsService.kt
package com.grateful.deadly.core.database

import android.util.Log
import org.json.JSONArray
import org.json.JSONObject
import java.net.HttpURLConnection
import java.net.URL
import java.util.UUID
import java.util.concurrent.CopyOnWriteArrayList
import javax.inject.Inject
import javax.inject.Named
import javax.inject.Singleton
import kotlinx.coroutines.*

/**
 * Fire-and-forget anonymous analytics client.
 * Buffers events in memory and flushes to the server periodically.
 */
@Singleton
class AnalyticsService @Inject constructor(
    private val appPreferences: AppPreferences,
    @Named("analyticsApiKey") private val apiKey: String,
    @Named("appVersionName") private val appVersion: String
) {
    companion object {
        private const val TAG = "AnalyticsService"
        private const val FLUSH_INTERVAL_MS = 30_000L
        private const val MAX_BUFFER_SIZE = 50
    }

    private val sessionId = UUID.randomUUID().toString()
    private val buffer = CopyOnWriteArrayList<AnalyticsEvent>()
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)

    init {
        startFlushTimer()
    }

    fun track(event: String, props: Map<String, Any> = emptyMap()) {
        if (!appPreferences.analyticsEnabled.value) return

        buffer.add(
            AnalyticsEvent(
                event = event,
                ts = System.currentTimeMillis(),
                iid = appPreferences.installId,
                sid = sessionId,
                platform = "android",
                appVersion = appVersion,
                props = props.ifEmpty { null }
            )
        )

        if (buffer.size >= MAX_BUFFER_SIZE) {
            flush()
        }
    }

    fun flush() {
        val events = mutableListOf<AnalyticsEvent>()
        // Drain buffer atomically
        while (buffer.isNotEmpty()) {
            val event = buffer.removeFirstOrNull() ?: break
            events.add(event)
        }
        if (events.isEmpty()) return

        scope.launch {
            try {
                postEvents(events)
            } catch (e: Exception) {
                Log.d(TAG, "Analytics flush failed (discarding): ${e.message}")
            }
        }
    }

    private fun postEvents(events: List<AnalyticsEvent>) {
        val baseUrl = appPreferences.apiBaseUrl
        val url = URL("$baseUrl/api/analytics")
        val connection = url.openConnection() as HttpURLConnection

        try {
            connection.requestMethod = "POST"
            connection.setRequestProperty("Content-Type", "application/json")
            connection.setRequestProperty("X-Analytics-Key", apiKey)
            connection.doOutput = true
            connection.connectTimeout = 10_000
            connection.readTimeout = 10_000

            val payload = JSONObject().apply {
                put("events", JSONArray().apply {
                    for (event in events) {
                        put(JSONObject().apply {
                            put("event", event.event)
                            put("ts", event.ts)
                            put("iid", event.iid)
                            put("sid", event.sid)
                            put("platform", event.platform)
                            put("app_version", event.appVersion)
                            if (event.props != null) {
                                put("props", JSONObject(event.props))
                            }
                        })
                    }
                })
            }

            connection.outputStream.use { os ->
                os.write(payload.toString().toByteArray(Charsets.UTF_8))
            }

            connection.responseCode // trigger the request
        } finally {
            connection.disconnect()
        }
    }

    private fun startFlushTimer() {
        scope.launch {
            while (isActive) {
                delay(FLUSH_INTERVAL_MS)
                flush()
            }
        }
    }

    private data class AnalyticsEvent(
        val event: String,
        val ts: Long,
        val iid: String,
        val sid: String,
        val platform: String,
        val appVersion: String,
        val props: Map<String, Any>?
    )
}

File: iosApp/deadly/Core/Service/AnalyticsService.swift
import Foundation
#if canImport(UIKit)
import UIKit
#endif

/// Fire-and-forget anonymous analytics client.
/// Buffers events in memory and flushes to the server periodically.
final class AnalyticsService: @unchecked Sendable {
    private let appPreferences: AppPreferences
    private let apiKey: String
    private let platform = "ios"
    private let sessionId = UUID().uuidString
    private let appVersion: String

    private let buffer = AnalyticsBuffer()
    private let flushInterval: TimeInterval = 30
    private let maxBufferSize = 50

    /// Computed each flush so changes to the dev custom-server setting take
    /// effect immediately without needing an app restart.
    private var baseURL: String { appPreferences.apiBaseUrl }

    private let flushQueue = DispatchQueue(label: "deadly.analytics.flush", qos: .utility)
    private var flushTimer: DispatchSourceTimer?

    init(appPreferences: AppPreferences, apiKey: String) {
        self.appPreferences = appPreferences
        self.apiKey = apiKey
        self.appVersion = Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String ?? "unknown"

        startFlushTimer()

        #if canImport(UIKit)
        NotificationCenter.default.addObserver(
            forName: UIApplication.willResignActiveNotification,
            object: nil,
            queue: .main
        ) { [weak self] _ in
            self?.flush()
        }
        #endif
    }

    // MARK: - Public API

    func track(_ event: String, props: [String: Any] = [:]) {
        guard appPreferences.analyticsEnabled else { return }

        let entry = AnalyticsEvent(
            event: event,
            ts: Int(Date().timeIntervalSince1970 * 1000),
            iid: appPreferences.installId,
            sid: sessionId,
            platform: platform,
            app_version: appVersion,
            props: props.isEmpty ? nil : props
        )

        buffer.append(entry)

        if buffer.count >= maxBufferSize {
            flush()
        }
    }

    func flush() {
        let events = buffer.drain()
        guard !events.isEmpty else { return }

        guard let request = buildPostRequest(for: events) else { return }

        // Fire-and-forget — silently discard failures
        URLSession.shared.dataTask(with: request).resume()
    }

    /// Force-flushes the buffer and reports success/failure via `completion`.
    /// On failure, events are restored to the buffer so they can be retried.
    /// Intended for the developer "Flush Analytics" tool.
    func flushNow(completion: @escaping (Bool, Int, String?) -> Void) {
        let events = buffer.drain()
        guard !events.isEmpty else {
            completion(true, 0, nil)
            return
        }

        guard let request = buildPostRequest(for: events) else {
            buffer.prepend(events)
            completion(false, events.count, "Invalid analytics URL")
            return
        }

        URLSession.shared.dataTask(with: request) { [weak self] _, response, error in
            if let error = error {
                self?.buffer.prepend(events)
                completion(false, events.count, error.localizedDescription)
                return
            }
            if let http = response as? HTTPURLResponse, !(200..<300).contains(http.statusCode) {
                self?.buffer.prepend(events)
                completion(false, events.count, "HTTP \(http.statusCode)")
                return
            }
            completion(true, events.count, nil)
        }.resume()
    }

    private func buildPostRequest(for events: [AnalyticsEvent]) -> URLRequest? {
        guard let url = URL(string: "\(baseURL)/api/analytics") else { return nil }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue(apiKey, forHTTPHeaderField: "X-Analytics-Key")

        let payload: [[String: Any]] = events.map { event in
            var dict: [String: Any] = [
                "event": event.event,
                "ts": event.ts,
                "iid": event.iid,
                "sid": event.sid,
                "platform": event.platform,
                "app_version": event.app_version,
            ]
            if let props = event.props {
                dict["props"] = props
            }
            return dict
        }

        let body: [String: Any] = ["events": payload]
        request.httpBody = try? JSONSerialization.data(withJSONObject: body)
        return request
    }

    // MARK: - Private

    private func startFlushTimer() {
        let timer = DispatchSource.makeTimerSource(queue: flushQueue)
        timer.schedule(deadline: .now() + flushInterval, repeating: flushInterval)
        timer.setEventHandler { [weak self] in
            self?.flush()
        }
        timer.resume()
        flushTimer = timer
    }
}

// MARK: - Thread-safe buffer

private final class AnalyticsBuffer: @unchecked Sendable {
    private let lock = NSLock()
    private var events: [AnalyticsEvent] = []

    var count: Int {
        lock.lock()
        defer { lock.unlock() }
        return events.count
    }

    func append(_ event: AnalyticsEvent) {
        lock.lock()
        defer { lock.unlock() }
        events.append(event)
    }

    func drain() -> [AnalyticsEvent] {
        lock.lock()
        defer { lock.unlock() }
        let drained = events
        events.removeAll()
        return drained
    }

    func prepend(_ restored: [AnalyticsEvent]) {
        lock.lock()
        defer { lock.unlock() }
        events.insert(contentsOf: restored, at: 0)
    }
}

// MARK: - Event model

private struct AnalyticsEvent {
    let event: String
    let ts: Int
    let iid: String
    let sid: String
    let platform: String
    let app_version: String
    let props: [String: Any]?
}
